### PR TITLE
ConcurrentGC Heap/Tenure Expand Fix

### PR DESCRIPTION
**Pull request by RSalman**
Fixes https://github.com/eclipse-openj9/openj9/issues/13206

This change switches the order between updating INIT table _(tuneToHeap call)_ and setting mark bits during expansion. This way, we guarantee that MM bits are set during heap expand when we miss to update INIT table (or it's not updated in time) given overlap of KO and Tenure Expand.

The issue being resolved is similar to what has been [outlined in a previous issue](https://github.com/eclipse-openj9/openj9/issues/8020#issuecomment-573843125), where MM Bits get missed for init. Here, we have another similar timing hole, when one thread kicks off concurrent, `Concurrent_OFF -> (Concurrent_INIT or INIT Complete)`, while the expanding thread is in middle of `heapReconfigured`.

With the original ordering of `heapReconfigured` - first, we attempt to set the mark bits during the expand with Concurrent ON. With Concurrent_OFF, we delay setting the bits and leave it for `Concurrent_INIT` phase (this requires update to the init table).  Hence, for Concurrent_OFF we forgo setting bits during expansion and update the init table, we will expect `tuneToHeap` to do `determineInitRanges` to update init ranges table. The issue occurs when we don't set mark bits during expansion (given concurrent_OFF) but concurrent starts in between after checking Concurrent State and prior to updating the init ranges. Here, we either don't update init table (since init is in progress) or miss to update it in time.

With these changes to `heapReconfigured`, we will try to set mark bits during expansion only after check to update init range table. With this, any state transitions resulting in init table not being updated (or being updated to late) will be caught and accounted for as bits will be set after.

Signed-off-by: Salman Rana <salman.rana@ibm.com>

### CODE REVIEW

File: gc/base/standard/ConcurrentGC.cpp [hunk 1]
```diff
@@ -3382,7 +3405,7 @@ MM_ConcurrentGC::heapReconfigured(MM_EnvironmentBase *env, HeapReconfigReason re
 	 */
 
 	if ((lowAddress != NULL) && (highAddress != NULL)) {
-		Assert_MM_true(_rebuildInitWorkForAdd && (reason == HEAP_RECONFIG_EXPAND));
```
Comment (RSalman) [inline gc/base/standard/ConcurrentGC.cpp:3385]: This condition is no longer valid - `_rebuildInitWorkForAdd` may be rest by tuneToHeap.

**Comment by RSalman** [comment]
fyi @amicic @dmitripivkine

**Comment by RSalman** [comment]
I'm still running some grinders and doing some testing on these changes.

File: gc/base/standard/ConcurrentGC.cpp [hunk 2]
```diff
@@ -3368,6 +3368,29 @@ MM_ConcurrentGC::heapReconfigured(MM_EnvironmentBase *env, HeapReconfigReason re
 {
 	Assert_MM_true(reason != HEAP_RECONFIG_NONE);
 
+	/* If called outside a global collection for a heap expand/contract.. */
+	if((reason == HEAP_RECONFIG_CONTRACT) || (reason == HEAP_RECONFIG_EXPAND)) {
+		Assert_MM_true(_rebuildInitWorkForAdd || _rebuildInitWorkForRemove);
+		if (!_stwCollectionInProgress) {
+			/* ... and a concurrent cycle has not yet started then we
+			 *  tune to heap here to reflect new heap size
+			 *  Note: CMVC 153167 : Under gencon, there is a timing hole where
```
Comment (dmitripivkine) [inline gc/base/standard/ConcurrentGC.cpp:3377]: I think we should not refer to internal issue in open

**Comment by RSalman** [comment]
This is somewhat non-trivial change, I've considered the implications closely. There are a few different state changes with heap expansion to consider and I've stepped through `tuneToHeap`/`adjustTraceTarget` to see if theres a restriction on ordering of the two. I've extended testing of the original issue to confirm. Overall, it should be a safe change. I'll confirm after more testing and looking over some broader code. 

One thing this introduces is that an expanded MM range may be inited twice in a very rare case, when init kicks off right after tuneToHeap but before the check to set bits in place. Here, init table will be updated for Concurrent_INIT and we'll also set the bits during expansion. I think its a fair compromise, I've considered other alternatives, any other solution will require more complex changes.

File: gc/base/standard/ConcurrentGC.cpp [hunk 3]
```diff
@@ -3368,6 +3368,29 @@ MM_ConcurrentGC::heapReconfigured(MM_EnvironmentBase *env, HeapReconfigReason re
 {
 	Assert_MM_true(reason != HEAP_RECONFIG_NONE);
 
+	/* If called outside a global collection for a heap expand/contract.. */
+	if((reason == HEAP_RECONFIG_CONTRACT) || (reason == HEAP_RECONFIG_EXPAND)) {
```
Comment (dmitripivkine) [inline gc/base/standard/ConcurrentGC.cpp:3372]: Would you please add missed space after `if`?

**Comment by dmitripivkine** [comment]
This change looks reasonable for me. Let's Aleks review it as well

**Comment by RSalman** [comment]
The code regarding ConcurrentGC expansion is somewhat complicated, tenure expansion with CS adds on top of that. I don't know if it's necessary but Just an idea, I think we can eventually have Concurrent Kickoff and Expansion mutually exclusive to simplify things.

**Comment by RSalman** [comment]
I've amended the original commit with a change in  `MM_ConcurrentGCIncrementalUpdate::heapAddRange`. 

This change is not related to the crash, but it's related to CM overlapping with heap expansion, it solves a potential issues that can arise with the card table. Like the mark map, expanded card table is inited (cleaned) during heap expansion when Concurrent is active at the time of expand, otherwise cards are cleaned during concurrent init. However, if Concurrent kicks off during expansion, after check to clean cards, then we can miss to update the init table or update it too late.

**Comment by RSalman** [comment]
With the card table change, we will preemptively clean new cards if we are in the middle of a CS cycle during expansion even if CM is off. CS cycle during tenure expand is the only way for such a case to occur, at worst, we will end up unnecessarily cleaning card during the expand with CS when CM is off. This is fine because cleaning cards (i.e `memset`) is negligible, each 512 bytes of expanded heap will require to `memset` of 1 byte... For every 100mb of expansion it's 2kb to `memset`

**Comment by RSalman** [comment]
>I'm wondering if we can have problem with opposite direction now: concurrent going from ON to OFF (exactly between init and direct MM clearing)

`ON -> OFF` transition can't happen during heap expand, transition to OFF state will always happen with Exclusive Access, during STW. 

All in all, the problematic  transitions during heap expand are: `OFF -> INIT RUNNING` and `OFF -> INIT RUNNING -> INIT COMPLETE`. Moving past `INIT COMPLETE` requires Exclusive as well.

File: gc/base/standard/ConcurrentGC.cpp [hunk 4]
```diff
@@ -3368,6 +3368,29 @@ MM_ConcurrentGC::heapReconfigured(MM_EnvironmentBase *env, HeapReconfigReason re
...
+			} else {
+				/* Heap expand/contract is during a concurrent cycle..we need to adjust the trace target so
+				 * that the trace rate is adjusted correctly on  subsequent allocates.
+				 */
+				adjustTraceTarget();
+			}
+		}
+	}
+
 	/* _rebuildInitWorkForAdd/_rebuildInitWorkForRemove are not sufficient for determining on how to proceed with headReconfigured
```
Comment (amicic) [inline gc/base/standard/ConcurrentGC.cpp:3394]: this comment should still be at top

**Comment by RSalman** [comment]
> Another possible high level approach to fixing this is taking a snapshot of _stats.getExecutionMode() in a local variable at the start of heapReconfigured(), so we know it does not change and we are guarantieed one or the other clearing path is taken.

I've considered this before and found that it won't resolve the issue. Using cached value to make decisions won't do much. for example, if we decide to to update int table based on cache value of OFF and INIT has actually started, then updating the table might still be done too late. and there can be other issues, like races when we update the table while it is being consumed by INIT.  

The other option is to cache the value at the start and recheck at the end to see if we went from OFF to ON, this will just be duplicating code code and the reorder is simpler.

**Comment by amicic** [comment]
> 
> 
> With the card table change, we will preemptively clean new cards during expansion if we are in the middle of a CS cycle, even if CM is off. CS cycle during tenure expand is the only way for such a case to occur, we will end up unnecessarily cleaning card during the expand with CS when CM is off. This is fine because cleaning cards (i.e `memset`) is negligible, each 512 bytes of expanded heap will require to `memset` of 1 byte... For every 100mb of expansion it's 2kb to `memset`. Also, I think it's a fair compromise as CS shouldn't be expanding Tenure too frequently.

Not the cleanest solution possible (not that I can suggest any better), but bottom line I'm not worried about extra work it would append (if we had to copy a large object and expand tenure for that, that alone is much bigger penalty for the mutator to start with).

But, add some comments based on these PR commetns, in the code itself.

**Review by amicic — APPROVED** [review]

**Review by dmitripivkine — APPROVED** [review]

**Comment by dmitripivkine** [comment]
@fjeremic Would you please proceed with this?

**Comment by fjeremic** [comment]
@RSalman is this ready to go? If so please remove the WIP from the PR and tag me in a comment and I will proceed with the merge.

**Comment by dmitripivkine** [comment]
> @RSalman is this ready to go? If so please remove the WIP from the PR and tag me in a comment and I will proceed with the merge.

Salman is on vacation. Would you please remove WIP prefix yourself (I think it was forgotten)? I believe it is ready - and has been approved by Aleks too.

**Comment by RSalman** [comment]
Thanks for merging and apologies for that, the WIP tag was forgotten!

**Pull request merged**